timeutil/schedule.go

```diff
@@ -233,3 +475,242 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+func parseWeekSpan(s string) (span WeekSpan, err error) {
+	var parsed WeekSpan
+
+	split := strings.Split(s, spanToken)
+	if len(split) > 2 {
+		return span, fmt.Errorf("cannot parse %q: invalid week span", s)
+	}
+
+	parsed.Start, err = parseWeekday(split[0])
+	if err != nil {
+		return span, fmt.Errorf("cannot parse %q: %q is not a valid weekday", s, split[0])
+	}
+
+	if len(split) == 2 {
+		parsed.End, err = parseWeekday(split[1])
+		if err != nil {
+			return span, fmt.Errorf("cannot parse %q: %q is not a valid weekday", s, split[1])
+		}
+	} else {
+		parsed.End = parsed.Start
+	}
+
+	if parsed.End.Pos < parsed.Start.Pos {
+		// eg. mon4-mon1
+		return span, fmt.Errorf("cannot parse %q: unsupported schedule", s)
+	}
+
+	if (parsed.Start.Pos != 0) != (parsed.End.Pos != 0) {
+		return span, fmt.Errorf("cannot parse %q: mixed weekday and nonweekday", s)
+	}
+
+	return parsed, nil
+}
+
+// parseClockSpan parses a time specification which can either be `<hh>:<mm>` or
+// `<hh>:<mm>[-~]<hh>:<mm>[/count]`. Alternatively the span can be one of
+// special tokens `-`, `~` (followed by an optional [/count]) that indicate a
+// whole day span, or a whole day span with spread respectively.
+func parseClockSpan(s string) (span ClockSpan, err error) {
+	var parsed ClockSpan
+	var rest string
+
+	// timespan = time ( "-" / "~" ) time [ "/" ( time / count ) ]
+
+	parsed.Split, rest, err = parseCount(s)
+	if err != nil {
+		return span, fmt.Errorf("cannot parse %q: not a valid interval", s)
+	}
+
+	if strings.Contains(rest, spreadToken) {
+		// timespan uses "~" to indicate that the actual event
+		// time is to be spread.
+		parsed.Spread = true
+		rest = strings.Replace(rest,
+			spreadToken,
+			spanToken, 1)
+	}
+
+	if strings.Contains(rest, spanToken) {
+		if rest == "-" {
+			// whole day span
+			parsed.Start = Clock{0, 0}
+			parsed.End = Clock{24, 0}
+		} else {
+			parsed.Start, parsed.End, err = parseClockRange(rest)
+		}
+	} else {
+		parsed.Start, err = ParseClock(rest)
+		if err == nil {
+			parsed.End = parsed.Start
+		}
+	}
+
+	if err != nil {
+		return span, fmt.Errorf("cannot parse %q: not a valid time", s)
+	}
+
+	return parsed, nil
+}
+
+// parseWeekday parses a single weekday (eg. wed, mon5),
+func parseWeekday(s string) (week Week, err error) {
+	l := len(s)
+	if l != 3 && l != 4 {
+		return week, fmt.Errorf("cannot parse %q: invalid format", s)
+	}
+
+	var day = s
+	var pos uint
+	if l == 4 {
+		day = s[0:3]
+		v, err := strconv.ParseUint(s[3:], 10, 32)
+		if err != nil || v < 1 || v > 5 {
+			return week, fmt.Errorf("cannot parse %q: invalid week number", s)
+		}
+		pos = uint(v)
+	}
+
+	weekday, ok := weekdayMap[day]
+	if !ok {
+		return week, fmt.Errorf("cannot parse %q: invalid weekday", s)
+	}
+
+	return Week{weekday, pos}, nil
+}
+
+// parseCount will parse the string containing a count token and return the
+// count, the rest of the string with count information removed or an error
```

Comment:
// ...
// count and the rest of the string with count information removed, or an error.